## Importing libraries and dataset

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import tensorflow as tf
from tensorflow import keras

In [3]:
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
print(train_images.shape,test_images.shape)

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 3us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 5s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0s/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step
(60000, 28, 28) (10000, 28, 28)


## Data Analysis

Reading the first 10 labels

In [4]:
print(train_labels[:10])

[9 0 0 3 0 2 7 2 5 5]


Determining the class names for each class

In [5]:
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat', 'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']
for i in range(10):
    print(class_names[train_labels[i]])

Ankle boot
T-shirt/top
T-shirt/top
Dress
T-shirt/top
Pullover
Sneaker
Pullover
Sandal
Sandal


Reading the first 10 image data consisting of the pixels of the image

In [ ]:
print(train_images[:10])

Let us also have a look at some of the images in the dataset to get an idea of what do we have to detect

In [ ]:
i=0
for image in train_images[:5]:
    plt.figure()
    plt.imshow(image,cmap="Greens")
    plt.colorbar()
    plt.title(f"Class: {class_names[train_labels[i]]}")
    plt.grid(False)
    plt.show()
    i+=1

Distribution of number of images: Here we would like to see the frequency of each of the image types and analyze their distribution

In [ ]:
sns.histplot(train_labels)
plt.xlabel("Types of images")
plt.xticks(ticks=range(len(class_names)),labels=class_names,rotation=45)

We observe that the images are equally distributed in the dataset i.e. we have equal number of images for each of the image types

## Data Preprocessing

Scaling the data: The train_images and test_images data consist of a wide range of numbers as they depict the pixels from 0-255. This leads to the nueral network attaching incorrect weights and biases to the features and reduces its accuracy. Thus we would scale the data to make sure that all the images data lies between 0 and 1.

In [ ]:
test_images = test_images / 255.0
train_images = train_images / 255.0
print(train_images[:10])
print(test_images[:10])

## Building the neural network model

We will build a sequential model consisting of Dense Layers. The initial layer is called as the input layer which has been flattened to contain 784 neurons. Then we have a hidden layer which uses the relu activation function consisting of 128 neuron and finally we have the output layer with 10 neurons corresponding to the number of classes of the image. The output layer uses the softmax activation function for multiclass classification.

In [23]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28,28)),
    keras.layers.Dense(128,activation='relu'),
    keras.layers.Dense(10,activation='softmax')
])
model.summary()

c:\Users\Areeb\Desktop\VS CODE Workspace\myenv\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten (Flatten)               │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       100,480 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 101,770 (397.54 KB)

 Trainable params: 101,770 (397.54 KB)

 Non-trainable params: 0 (0.00 B)

Now we compile the model which we have just built. We will use the Adam optimizer function which is used to calculate the derivatives of the weights and biases in the neural network more efficiently during back propagation as this algorithm keeps on updating the learning rate. We have further used the 'sparse_categorical_crossentropy' as our loss function and the 'accuracy' metric to measure the performance of the model.

In [31]:
model.compile(optimizer='adam',loss='sparse_categorical_crossentropy',metrics=['accuracy'])

We then fit the neural network to our training data and observe the performance of our model. We have used 10 epochs to train the model.

In [32]:
model.fit(train_images,train_labels,epochs=10)

Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 3s 1ms/step - accuracy: 0.7861 - loss: 0.6292
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8612 - loss: 0.3813
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8765 - loss: 0.3339
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8869 - loss: 0.3139
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8909 - loss: 0.2964
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8972 - loss: 0.2769
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9021 - loss: 0.2658
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9056 - loss: 0.2546
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9094 - loss: 0.2398
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.9103 - loss: 0.2390


We observe that the model shows a high accuracy of 91%. Lets now evaluate the performance of our model when it sees previously unknown data. We will use the testing data for this purpose.

In [33]:
loss,accuracy=model.evaluate(test_images,test_labels)
print(f"Accuracy: {accuracy*100}%")

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8831 - loss: 0.3478
Accuracy: 88.31999897956848%


The model shows an accuracy of 88.32% on the testing data. Some overfitting can be interpreted from this information but we can still conclude that our model generalizes pretty well to any kind of given data. 
Once we have built, trained and evaluated our model, we can now use this model to make some predictions regarding any images from the testing data.

In [ ]:
predictions = model.predict(test_images)

313/313 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
9 9


In [54]:
def verify_image(index):
    print(f"Class: {class_names[test_labels[index]]} Predicted: {class_names[np.argmax(predictions[index])]}")
    plt.figure()
    plt.imshow(test_images[index],cmap="Grays")
    plt.colorbar()
    plt.title("Actual image")
    plt.grid(False)
    plt.show()

In [ ]:
index=int(input("Enter the index of the image you want to verify: "))
verify_image(index)

## Conclusion

We have successfully built a neural network which takes image data consisting of pixels of the images as input and processes the data through various layers to finally produce the given output. A sufficient accuracy score was achieved in detecting the correct and appropriate class of the given images. The output layer produces predictions generating the possibilities of the image belonging to each of the classes and then we pick up the class with the highest probability to make our prediction. 